src/query/function_score_query.rs

```diff
@@ -0,0 +1,289 @@
+use crate::query::{EnableScoring, Explanation, Query, Scorer, Weight};
+use crate::{DocId, DocSet, Score, SegmentReader, TantivyError};
+use std::fmt;
+use std::sync::Arc;
+
+type Function = Arc<dyn Fn(&SegmentReader, Score, DocId) -> Score + Send + Sync + 'static>;
```

Comment:
If this closure is supposed to access the `SegmentReader`, I think it should return `crate::Result<Score>` to be able to forward errors. But then again `Scorer::score` currently does not allow errors so we cannot forward this anywhere.

Not sure how to best handle this, expect for making all scoring fallible, but that might have a measurable performance cost?

Reply:
We can handle the fallback case like this:

```rust
type Function = Arc<dyn Fn(&SegmentReader, Score, DocId) -> crate::Result<Score> + Send + Sync + 'static>;

impl<TDocSet: DocSet + 'static> Scorer for FunctionScorer<TDocSet> {
    fn score(&mut self) -> Score {
        let score = self.scorer.score();
        let doc_id = self.doc();
        match (self.function)(self.segment_reader.as_ref(), score, doc_id) {
            Ok(new_score) => new_score,
            Err(_) => score,
        }
    }
}
```

But then we should have a mechanism to tell the users that some error happens. Do you have any idea on this?

Reply:
> But then we should have a mechanism to tell the users that some error happens. Do you have any idea on this?

I would refrain from silently not adjusting the score. Either the operations work or they don't. If errors are supported, they should be propagated.

Just falling back to not modifying the score can also be implemented in the given function itself without any support in the scorer if that is indeed the chosen approach in a specific situation.

Reply:
I think the error handling will be resolved by https://github.com/quickwit-oss/tantivy/pull/2396#discussion_r1598176436 as a side effect.